### PR TITLE
migrations: Fix migration discrepancy on a system table.

### LINKED ISSUE #77980: system.public.statement_diagnostics_requests table definition has discrepancy after upgrading from a old version
When we upgrade CRDB binary version running in nodes in a cluster, migration logic will be triggered to make sure tables in system database are "consistent" (i.e. same definition) as that of a newly bootstrapped cluster.

A recent roachtest showed there is discrepancy in the system.public.statement_diagnostics_requests table. Namely, the actual return of `SHOW CREATE TABLE` after upgrading is
```
CREATE TABLE public.statement_diagnostics_requests (	
	id INT8 NOT NULL DEFAULT unique_rowid(),	
	completed BOOL NOT NULL DEFAULT false,	
	statement_fingerprint STRING NOT NULL,	
	statement_diagnostics_id INT8 NULL,	
	requested_at TIMESTAMPTZ NOT NULL,	
	min_execution_latency INTERVAL NULL,	
	expires_at TIMESTAMPTZ NULL,	
	CONSTRAINT "primary" PRIMARY KEY (id ASC),	
	INDEX completed_idx (completed ASC, id ASC) STORING (statement_fingerprint),	
	INDEX completed_idx_v2 (completed ASC, id ASC) STORING (statement_fingerprint, min_execution_latency, expires_at),	
	FAMILY "primary" (id, completed, statement_fingerprint, statement_diagnostics_id, requested_at),	
	FAMILY fam_6_min_execution_latency (min_execution_latency),	
	FAMILY fam_7_expires_at (expires_at)	
);
```

while the expected output (as hard-coded in systemschema) is

```
CREATE TABLE public.statement_diagnostics_requests (
	id INT8 NOT NULL DEFAULT unique_rowid(),
	completed BOOL NOT NULL DEFAULT false,
	statement_fingerprint STRING NOT NULL,
	statement_diagnostics_id INT8 NULL,
	requested_at TIMESTAMPTZ NOT NULL,
	min_execution_latency INTERVAL NULL,
	expires_at TIMESTAMPTZ NULL,
	CONSTRAINT "primary" PRIMARY KEY (id ASC),
	INDEX completed_idx_v2 (completed ASC, id ASC) STORING (statement_fingerprint, min_execution_latency, expires_at)
```

Two things need to be done to close this issue:
1. Migration logic should drop the old index (`completed_idx`)
2. Investigate and make sure the three newly added column do not have a family of its own. Instead, they should be placed into the existing family.

Similar issue: #77979 

Jira issue: CRDB-13881

### PR DESCRIPTION
**Pull request by Xiang-Gu**
Previously, the migration code for system table
`statement_diagnostics_requests` results in discrepancies when compared
to the table definition after bootstrapping. The discrepancies are
summarized in #77980.

This PR fixes it to ensure this table has exactly the same definition
after cluster upgrading as that from bootstrapping.

Release note: None

Fixes: #77980

Sister Issue: #78302

Related PR that discovers this discrepancy: #77970

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
Wonderful job catching this and great job fixing it!

Can you extend `TestAlterSystemStmtDiagReqs` to also check that the index no longer exists and the families look correct?
Can you associate the issue with the roachtest you're writing in some form? 

:lgtm: mod the testing requests and commit message feedback.

I added the `backport-22.1` label which will result in a backport being created automatically once this merges. 

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-:-MzHhjLsDP5tejhLREMX:b-regav0)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[-- commits, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-MzHhu3l4Q7sDzHtgMRZ):*
I don't think we need a release note because we're fixing a bug which was not shipped in a release. 



_Code quote:_
```
  Release note (bug fix): Ensure the migration logic for
  `system.statement_diagnostics_requests` results in a table definition
  that is identical to that from bootstrapping.
```
___
*[-- commits, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-MzHhvUwEJmfTka6n34_):*
The backport to 22.1 will fix this issue. I find it can be problematic to close the issue before the backport.

**Review by yuzefovich — COMMENTED** [review]
Thanks for catching this!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-:-MzIaTqaAbP9yljwcn8F:b-x3gtbh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @Xiang-Gu)

___
*[-- commits, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-MzHhvUwEJmfTka6n34_):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The backport to 22.1 will fix this issue. I find it can be problematic to close the issue before the backport. 
</blockquote></details>

My preferred way to go about it is to not reference the issue in the commit message, but add the necessary details to the PR description only. This way you can add the `Fixes #issue-number` part to the relevant PR (the one against 22.1 branch in this case). An additional benefit to the cleanliness is that the issue itself won't be mentioned every time you amend the commit and force-push the branch.

**Review by Xiang-Gu — COMMENTED** [review]
Thanks. I've extended the test to check that the intended-to-delete index is indeed deleted. The families check is already included in the `primary` family check.

Also, I'll associate this pr with the roachtest I'm working on.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-:-MzM0J4Q4IdttUyuOzpf:bz41558)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @yuzefovich)

___
*[-- commits, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-MzHhu3l4Q7sDzHtgMRZ):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think we need a release note because we're fixing a bug which was not shipped in a release. 
</blockquote></details>

Done.
___
*[-- commits, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-MzHhvUwEJmfTka6n34_):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

My preferred way to go about it is to not reference the issue in the commit message, but add the necessary details to the PR description only. This way you can add the `Fixes #issue-number` part to the relevant PR (the one against 22.1 branch in this case). An additional benefit to the cleanliness is that the issue itself won't be mentioned every time you amend the commit and force-push the branch.
</blockquote></details>

Thanks for the suggestion. I'll remove the issue reference from commit message but add them back in the PR.

**Review by ajwerner — APPROVED** [review]
:lgtm: once you defeat the linter 

> Reviewed 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-:-MzMW1a7AiIIWhNcAuaP:b-rfqe1n)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @Xiang-Gu, and @yuzefovich)

___
*[pkg/migration/migrations/schema_changes.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-MzMVp8TD9hYtuIq0zHv:-MzMVp8U6M-tcuY7Fjo1:b-vhdj6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ced98776e869ccead531f50e74d4341405f1b8e/pkg/migration/migrations/schema_changes.go#L260)):*
> ```Go
> 	} else {
> 		return false, nil
> 	}
> ```

the linter is mad at this, how about writing this like:

```go
idx, _ := storedTable.FindIndexWithName(indexName)
return idx == nil, nil
```



_Code quote:_
```Go
	if _, err := storedTable.FindIndexWithName(indexName); err != nil {
		return true, nil
	} else {
		return false, nil
	}
```

**Review by Xiang-Gu — COMMENTED** [review]
Done. Thanks

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-:-MzMsb8G1KxTVWGqgxFK:bnputky)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @yuzefovich)

___
*[pkg/migration/migrations/schema_changes.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78679#-MzMVp8TD9hYtuIq0zHv:-MzMsZgPEA6_Qdxi3xo3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ced98776e869ccead531f50e74d4341405f1b8e/pkg/migration/migrations/schema_changes.go#L260)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

the linter is mad at this, how about writing this like:

```go
idx, _ := storedTable.FindIndexWithName(indexName)
return idx == nil, nil
```
</blockquote></details>

Done

**Pull request merged**
